### PR TITLE
(BREAKING) Adopt OpenMRS' global Expression Runner

### PR DESCRIPTION
**Pull request by samuelmale**
## Requirements

- [X] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [X] My work includes tests or is validated by existing tests.

## Summary

Migrates away from using the custom expression runner that had compilation issues to using OpenMRS' new expression runner which is a safer option. This change should fix most, if not all, of the expression compilation-related issues. 

Screenshot 2024-10-10 at 02 53 04

**Potential breaking changes**:
- Async Expressions
Previously, the syntax allowed for expressions like `"resolve(getAsyncValue()) > someOtherField"`, which returned a Promise that resolved to a boolean value. The `resolve()` function was used to handle asynchronous fragments within the expression.
Now, the expression runner supports the Promise and then pattern directly. To achieve similar functionality, you must now use syntax like `"getAsyncValue().then(value => value > someOtherField)"`. 

- Variable Argument Handling  
OpenMRS' global expression evaluator only supports variable arguments; it doesn't attempt inferring string arguments. Previously, you could call a function like `calcMonthsOnART('artStartDate')`, where the string `'artStartDate'` would be resolved into a variable if a corresponding field ID existed. This meant that the engine would interpret the string as a reference to an actual variable. With the new expression evaluator, you are expected to use variable arguments: `calcMonthsOnART(artStartDate)`

## Related Issue

- https://openmrs.atlassian.net/browse/O3-4075
- https://openmrs.atlassian.net/browse/O3-3337

### CODE REVIEW

File: __mocks__/forms/rfe-forms/bmi-test-form.json [hunk 1]
```diff
@@ -1,69 +1,69 @@
...
+              "validators": [],
+              "id": "weight"
+            },
+            {
+              "label": "BMI",
+              "type": "obs",
+              "questionOptions": {
+                "rendering": "text",
+                "calculate": {
+                  "calculateExpression": "calcBMI(height, weight)"
```
Comment (samuelmale) [inline __mocks__/forms/rfe-forms/bmi-test-form.json:47]: Key change: `calcBMI('height', 'weight')` -> `calcBMI(height, weight)`

File: __mocks__/forms/rfe-forms/bsa-test-form.json [hunk 2]
```diff
@@ -1,69 +1,69 @@
...
+              "validators": [],
+              "id": "weight"
+            },
+            {
+              "label": "BSA",
+              "type": "obs",
+              "questionOptions": {
+                "rendering": "text",
+                "calculate": {
+                  "calculateExpression": "calcBSA(height, weight)"
```
Comment (samuelmale) [inline __mocks__/forms/rfe-forms/bsa-test-form.json:47]: Key change: `"calcBSA('height','weight')"` -> `"calcBSA(height, weight)"`

File: __mocks__/forms/rfe-forms/edd-test-form.json [hunk 3]
```diff
@@ -1,88 +1,88 @@
...
-                            "validators": [],
-							"id": "LMP"
-						},
-						{
-							"label": "EDD",
-							"type": "obs",
-							"questionOptions": {
-								"rendering": "date",
-                                "calculate": {
-                                    "calculateExpression": "calcEDD('LMP')"
```
Comment (samuelmale) [inline __mocks__/forms/rfe-forms/edd-test-form.json:43]: Key change

File: __mocks__/forms/rfe-forms/external_data_source_form.json [hunk 4]
```diff
@@ -1,40 +1,39 @@
...
+          "isExpanded": "true",
+          "questions": [
+            {
+              "label": "Body Weight",
+              "type": "obs",
+              "questionOptions": {
+                "rendering": "number",
+                "concept": "560555AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
+                "calculate": {
+                  "calculateExpression": "api.getLatestObs('50512033-e047-4855-b2d3-1a6d9e889ff4', '560555AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA').then(obs => obs.valueNumeric)"
```
Comment (samuelmale) [inline __mocks__/forms/rfe-forms/external_data_source_form.json:18]: Key change (async expression syntax)

**Comment by samuelmale** [comment]
FYI @michaelbontyes @gracepotma

**Comment by denniskigen** [comment]
FYI @jnsereko

**Review by ibacher — APPROVED** [review]
LGTM

**Review by pirupius — COMMENTED** [review]
Thanks @samuelmale for this. Had u tested it out with the other formula already?

**Pull request merged**